File: bindings/python/insight/signal/wavelets.py
"""Wavelet functions for signal processing.

Provides Morlet and Ricker wavelet generators, and the continuous wavelet
transform (CWT).
"""

from insight._insight import signal as _signal

__all__ = [
    "morlet",
    "ricker",
    "morlet2",
    "cwt",
]


def morlet(M, w=5.0, s=1.0, complete=True):
    """Return a complex Morlet wavelet.

    Args:
        M: Length of the wavelet in samples.
        w: Omega0 (center frequency).  Default is 5.0.
        s: Scaling factor.  Default is 1.0.
        complete: Whether to use the complete (with correction term) or
            standard version.  Default is True.

    Returns:
        Complex Morlet wavelet array of length *M*.
    """
    return _signal.morlet(M, w=w, s=s, complete=complete)


def ricker(points, a):
    """Return a Ricker wavelet (Mexican hat wavelet).

    Args:
        points: Number of points in the wavelet.
        a: Width parameter of the wavelet.

    Returns:
        Ricker wavelet array of length *points*.
    """
    return _signal.ricker(points, a)


def morlet2(M, s=1.0, w=5.0):
    """Return a complex Morlet wavelet (alternative parameterization).

    Args:
        M: Length of the wavelet in samples.
        s: Scaling factor.  Default is 1.0.
        w: Omega0 (center frequency).  Default is 5.0.

    Returns:
        Complex Morlet wavelet array of length *M*.
    """
    return _signal.morlet2(M, s=s, w=w)


def cwt(data, wavelet, widths):
    """Compute the continuous wavelet transform.

    Args:
        data: Input signal array (1-D).
        wavelet: Wavelet function.  Should accept ``(points, width)``
            and return a 1-D wavelet array.
        widths: Sequence of widths (scales) at which to compute the CWT.

    Returns:
        2-D array of shape ``(len(widths), len(data))`` containing the
        CWT coefficients.
    """
    return _signal.cwt(data, wavelet, list(widths))

File: bindings/lua/insight/signal/wavelets.lua
--- Wavelet functions.
-- Provides Morlet, Ricker (Mexican hat), and Morlet2 wavelet functions.
-- @module insight.signal.wavelets

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

--- Compute a Morlet wavelet.
-- @int M Length of the wavelet (number of points).
-- @number[opt=5.0] w Frequency parameter (omega0).
-- @number[opt=1.0] s Scaling factor.
-- @bool[opt=true] complete Whether to use the complete version.
-- @treturn Array Complex Morlet wavelet.
M.morlet = _wrap({ "M", "w", "s", "complete" }, function(M_val, w, s, complete)
  return sig.morlet(M_val, w, s, complete)
end)

--- Compute a Ricker (Mexican hat) wavelet.
-- @int points Number of points.
-- @number a Width parameter.
-- @treturn Array Ricker wavelet.
M.ricker = _wrap({ "points", "a" }, function(points, a)
  return sig.ricker(points, a)
end)

--- Compute a Morlet2 wavelet.
-- @int M Length of the wavelet (number of points).
-- @number s Scale parameter.
-- @number[opt=5.0] w Frequency parameter (omega0).
-- @treturn Array Complex Morlet2 wavelet.
M.morlet2 = _wrap({ "M", "s", "w" }, function(M_val, s, w)
  return sig.morlet2(M_val, s, w)
end)

return M
